README.md

# LORECOVER <img src="man/figures/lakeO.png" align="right" alt="" width="120" />

[![DOI](https://zenodo.org/badge/349568854.svg)](https://zenodo.org/badge/latestdoi/349568854)

## Table of Contents

-   [Introduction](#intro)
-   [Citing package](#cite)
-   [Installation](#install)

## Introduction <a name="intro"></a>

Functions in this package calculate normal and recovery stage envelopes
for Lake Okeechobee. Complete documentation for this performance measure
can be found on the
[USACE](https://usace.contentdm.oclc.org/utils/getfile/collection/p16021coll7/id/14096)
webpage.

<div class="figure" style="text-align: center">

<img src="man/figures/README-unnamed-chunk-1-1.png" alt="Generalized normal (left) and recovery (right) stage envelope scores."  />
<p class="caption">
Generalized normal (left) and recovery (right) stage envelope scores.
</p>

</div>

------------------------------------------------------------------------

``` r
library(AnalystHelper)
library(lubridate)
library(LORECOVER)

# Data from SFWMD
dat=DBHYDRO_daily(date.fun("2016-01-01"),date.fun("2021-12-31"),15611)
dat$CY=as.numeric(format(dat$Date,"%Y"))
```

#### Normal Lake Stage Envelope

``` r
#Normal Envelope
norm=subset(dat,CY==2016)
norm.val=norm_env(norm)
```

<div class="figure" style="text-align: center">

<img src="man/figures/README-unnamed-chunk-4-1.png" alt="The Normal lake stage envelope and approximate corresponding scores that apply for lake stages outside the desired range. Scores are actually applied by the hundredths of a foot on a daily basis and may not correspond exactly to the boxes shown. For reference and as an example, 2016 stages are overlaid onto the envelope and corresponding scores are shown in the lower panel."  />
<p class="caption">
The Normal lake stage envelope and approximate corresponding scores that
apply for lake stages outside the desired range. Scores are actually
applied by the hundredths of a foot on a daily basis and may not
correspond exactly to the boxes shown. For reference and as an example,
2016 stages are overlaid onto the envelope and corresponding scores are
shown in the lower panel.
</p>

</div>

#### Recovery Lake Stage Envelope

``` r
#Recovery Envelope
rec=subset(dat,CY==2019)
rec.val=rec_env(rec)
```

<div class="figure" style="text-align: center">

<img src="man/figures/README-unnamed-chunk-6-1.png" alt="The Recovery lake stage envelope and approximate corresponding scores that apply for lake stages outside the desired range. Scores are actually applied by the hundredths of a foot on a daily basis and may not correspond exactly to the boxes shown. For reference and as an example, 2019 stages are overlaid onto the envelope and corresponding scores are shown in the lower panel."  />
<p class="caption">
The Recovery lake stage envelope and approximate corresponding scores
that apply for lake stages outside the desired range. Scores are
actually applied by the hundredths of a foot on a daily basis and may
not correspond exactly to the boxes shown. For reference and as an
example, 2019 stages are overlaid onto the envelope and corresponding
scores are shown in the lower panel.
</p>

</div>

## Citing package <a name="cite"></a>

``` r
citation('LORECOVER')
#> 
#> To cite package 'LORECOVER' in publications use:
#> 
#>   Paul Julian (2021). LORECOVER: Lake Okeechobee Stage Envelope
#>   Performance Measure Calculation. R package version 0.2.2.
#>   https://github.com/SwampThingPaul/LORECOVER
#> 
#> A BibTeX entry for LaTeX users is
#> 
#>   @Manual{,
#>     title = {LORECOVER: Lake Okeechobee Stage Envelope Performance Measure Calculation},
#>     author = {Paul Julian},
#>     year = {2021},
#>     note = {R package version 0.2.2},
#>     url = {https://github.com/SwampThingPaul/LORECOVER},
#>   }
```

## Installation <a name="install"></a>

Development version can be installed from this repo.

``` r
install.packages("devtools");# if you do not have it installed on your PC
devtools::install_github("SwampThingPaul/LORECOVER")
```

------------------------------------------------------------------------


NEWS.md
# LORECOVER 0.1.0

* Added a `NEWS.md` file to track changes to the package.

* Package committed to Github

# LORECOVER 0.2.2

* Discovered some errors in the calculation of stage scores. Debugging complete.


R/rec_env.R
#' Recovery Lake Stage Envelope Score
#'
#'
#' @description Calculates a recovery lake stage envelope score
#' @param stg.data see details
#' @param allinfo TRUE/FALSE if TRUE, the function will return values used in computing score
#' @details
#' The input `stg.data` is a `data.frame` with columns:
#' \itemize{
#' \item{`Date` (as a `POSIXct` or `Date` variable)}
#' \item{`Data.Value` as stage elevation data in feet (NGVD29)}
#' }
#'
#' @return Returns a `data.frame` of original data and normal stage elevation score.
#'
#' @export
#'
#' @examples
#' # Example dataset (not real data)
#' dates=seq(as.Date("2016-01-01"),as.Date("2016-02-02"),"1 days")
#' dat=data.frame(Date=dates,Data.Value=runif(33,12,18))
#'
#' score=rec_env(dat)
#' # END

rec_env=function(stg.data,allinfo=FALSE){
  stg.data$month=as.numeric(format(stg.data$Date,"%m"))
  stg.data$CY=as.numeric(format(stg.data$Date,"%Y"))
  stg.data$day=as.numeric(format(stg.data$Date,"%d"))

  rec.score=data.frame(rec.stage.score)
  stg.data=merge(stg.data[,c("Date","month","day","CY","Data.Value")],
                 rec.score,c("month","day"),all.x=T)
  stg.data=stg.data[order(stg.data$Date),]

  # leap=leap_year(stg.data$Date)
  # yr=unique(stg.data[leap==T,"CY"])
  # vars=c("LP3", "LP2.5", "LP2", "LP1.5", "LP1", "LP0.5","UP0.5", "UP1", "UP2")
  # if(length(yr)>0){
  #   # Fills values for leap year
  # for(i in 1:length(yr)){
  # # leap_dates=seq(as.Date(paste(yr[i],02,28,sep="-")),as.Date(paste(yr[i],03,01,sep="-")),"1 days")
  # # stg.data[stg.data$CY==yr[i]&as.Date(stg.data$Date)%in%leap_dates,vars]=na.approx(stg.data[stg.data$CY==yr[i]&as.Date(stg.data$Date)%in%leap_dates,vars])
  #   leap_dates=as.Date(paste(yr[i],02,29,sep="-"))
  #   stg.data[stg.data$CY==yr[i]&as.Date(stg.data$Date)%in%leap_dates,vars]=rec.score[rec.score$month==2&rec.score$day==28,vars]
  #
  # }
  # }
  ##
  zs=data.frame(zone=c("LP3", "LP2.5", "LP2", "LP1.5", "LP1", "LP0.5","UP0.5", "UP1", "UP2"),
                score=c(3,2.5,2,1.5,1.0,0.5,0.5,1.0,2.0))
  zs.f=zs$zone
  zs.s=zs$score

  ##
  stg.data$zone=with(stg.data,ifelse(is.na(Data.Value)==T,0,
                                     ifelse(Data.Value<LP3,0,
                                            ifelse(Data.Value<LP2.5,1,
                                                   ifelse(Data.Value<LP2,2,
                                                          ifelse(Data.Value<LP1.5,3,
                                                                 ifelse(Data.Value<LP1,4,
                                                                        ifelse(Data.Value<LP0.5,5,
                                                                               ifelse(Data.Value<UP0.5,6,
                                                                                      ifelse(Data.Value<UP1,7,
                                                                                             ifelse(Data.Value<UP2,8,9)))))))))))
  stg.data$score1=with(stg.data,ifelse(zone==1,LP3,
                                       ifelse(zone==2,LP2.5,
                                              ifelse(zone==3,LP2,
                                                     ifelse(zone==4,LP1.5,
                                                            ifelse(zone==5,LP1,
                                                                   ifelse(zone==6,LP0.5,
                                                                          ifelse(zone==7,UP0.5,
                                                                                 ifelse(zone==8,UP1,NA)))))))))

  stg.data$score2=with(stg.data,ifelse(zone==1,LP2.5,
                                       ifelse(zone==2,LP2,
                                              ifelse(zone==3,LP1.5,
                                                     ifelse(zone==4,LP1,
                                                            ifelse(zone==5,LP0.5,
                                                                   ifelse(zone==6,UP0.5,
                                                                          ifelse(zone==7,UP1,
                                                                                 ifelse(zone==8,UP2,NA)))))))))
  stg.data$penalty=with(stg.data,
                        ifelse(zone==0,zs.s[1]+2*(LP3-Data.Value),
                               ifelse(zone==9,zs.s[9]+2*(Data.Value-UP2),
                                      ifelse(zone==6,0,
                                             zs.s[ifelse(zone==0,1,zone)]+(zs.s[zone+1]-zs.s[ifelse(zone==0,1,zone)])*(Data.Value-score1)/(score2-score1))))*sign(Data.Value-LP0.5))

  if(allinfo==TRUE){
    rslt=stg.data[,c("Date","Data.Value","zone","LP0.5","score1","score2","penalty")]}else{
      rslt=stg.data[,c("Date","Data.Value","penalty")]
    }
  # options(warn = -1)
  return(rslt)

}


data-raw/data_proc.R
library(openxlsx)

# Normal Stage Envelope
norm.stage.score=read.xlsx("C:/Julian_LaCie/SCCF/RECOVER/Lake Okeechobee/fromLOSOM.xlsx",sheet=1,startRow=1 )
# norm.stage.score$Date=date.fun(convertToDate(norm.stage.score$Date))

## CleanUp
# norm.stage.score$month=as.numeric(format(norm.stage.score$Date,"%m"))
# norm.stage.score$day=as.numeric(format(norm.stage.score$Date,"%d"))
# norm.stage.score=norm.stage.score[,c(11,12,2:10)]


# Recovery Stage Envelope
rec.stage.score=read.xlsx("C:/Julian_LaCie/SCCF/RECOVER/Lake Okeechobee/fromLOSOM.xlsx",sheet=2,startRow=1 )
# rec.stage.score$Date=date.fun(convertToDate(rec.stage.score$Date))

## CleanUp
# rec.stage.score$month=as.numeric(format(rec.stage.score$Date,"%m"))
# rec.stage.score$day=as.numeric(format(rec.stage.score$Date,"%d"))
# rec.stage.score=rec.stage.score[,c(11,12,2:10)]

usethis::use_data(norm.stage.score,internal=F,overwrite=T)
usethis::use_data(rec.stage.score,internal=F,overwrite=T)


R/norm_env.R
#' Normal Lake Stage Envelope Score
#'
#' @description Calculates a normal lake stage envelope score
#' @param stg.data see details
#' @param allinfo TRUE/FALSE if TRUE, the function will return values used in computing score
#'
#' @details
#' The input `stg.data` is a `data.frame` with columns:
#' \itemize{
#' \item{`Date` (as a `POSIXct` or `Date` variable)}
#' \item{`Data.Value` as stage elevation data in feet (NGVD29)}
#' }
#'
#' @return Returns a `data.frame` of original data and normal stage elevation score.
#'
#' @export
#'
#' @examples
#' # Example dataset (not real data)
#' dates=seq(as.Date("2016-01-01"),as.Date("2016-02-02"),"1 days")
#' dat=data.frame(Date=dates,Data.Value=runif(33,12,18))
#'
#' score=norm_env(dat)
#' # END

norm_env=function(stg.data,allinfo=FALSE){
  stg.data$month=as.numeric(format(stg.data$Date,"%m"))
  stg.data$CY=as.numeric(format(stg.data$Date,"%Y"))
  stg.data$day=as.numeric(format(stg.data$Date,"%d"))

  norm.score=data.frame(norm.stage.score)
  stg.data=merge(stg.data[,c("Date","month","day","CY","Data.Value")],
                 norm.score,c("month","day"),all.x=T)
  stg.data=stg.data[order(stg.data$Date),]

  # leap=leap_year(stg.data$Date)
  # yr=unique(stg.data[leap==T,"CY"])
  # vars=c("LP3", "LP2.5", "LP2", "LP1.5", "LP1", "LP0.5","UP0.5", "UP1", "UP2")
  # if(length(yr)>0){
  # # Fills values for leap year
  # for(i in 1:length(yr)){
  # # leap_dates=seq(as.Date(paste(yr[i],02,28,sep="-")),as.Date(paste(yr[i],03,01,sep="-")),"1 days")
  # # stg.data[stg.data$CY==yr[i]&as.Date(stg.data$Date)%in%leap_dates,vars]=na.approx(stg.data[stg.data$CY==yr[i]&as.Date(stg.data$Date)%in%leap_dates,vars])
  #
  # leap_dates=as.Date(paste(yr[i],02,29,sep="-"))
  # stg.data[stg.data$CY==yr[i]&as.Date(stg.data$Date)%in%leap_dates,vars]=norm.score[norm.score$month==2&norm.score$day==28,vars]
  #
  # }
  # }
  ##
  zs=data.frame(zone=c("LP3", "LP2.5", "LP2", "LP1.5", "LP1", "LP0.5","UP0.5", "UP1", "UP2"),
                score=c(3,2.5,2,1.5,1.0,0.5,0.5,1.0,2.0))
  zs.f=zs$zone
  zs.s=zs$score

  ##
  stg.data$zone=with(stg.data,ifelse(is.na(Data.Value)==T,0,
                                     ifelse(Data.Value<LP3,0,
                                            ifelse(Data.Value<LP2.5,1,
                                                   ifelse(Data.Value<LP2,2,
                                                          ifelse(Data.Value<LP1.5,3,
                                                                 ifelse(Data.Value<LP1,4,
                                                                        ifelse(Data.Value<LP0.5,5,
                                                                               ifelse(Data.Value<UP0.5,6,
                                                                                      ifelse(Data.Value<UP1,7,
                                                                                             ifelse(Data.Value<UP2,8,9)))))))))))
  stg.data$score1=with(stg.data,ifelse(zone==1,LP3,
                                       ifelse(zone==2,LP2.5,
                                              ifelse(zone==3,LP2,
                                                     ifelse(zone==4,LP1.5,
                                                            ifelse(zone==5,LP1,
                                                                   ifelse(zone==6,LP0.5,
                                                                          ifelse(zone==7,UP0.5,
                                                                                 ifelse(zone==8,UP1,0)))))))))

  stg.data$score2=with(stg.data,ifelse(zone==1,LP2.5,
                                       ifelse(zone==2,LP2,
                                              ifelse(zone==3,LP1.5,
                                                     ifelse(zone==4,LP1,
                                                            ifelse(zone==5,LP0.5,
                                                                   ifelse(zone==6,UP0.5,
                                                                          ifelse(zone==7,UP1,
                                                                                ifelse(zone==8,UP2,0)))))))))


  stg.data$penalty=with(stg.data,
                        ifelse(zone==0,zs.s[1]+2*(LP3-Data.Value),
                               ifelse(zone==9,zs.s[9]+2*(Data.Value-UP2),
                                      ifelse(zone==6,0,
                                             zs.s[ifelse(zone==0,1,zone)]+(zs.s[ifelse(zone==0,1,zone)+1]-zs.s[ifelse(zone==0,1,zone)])*(Data.Value-score1)/(score2-score1))))*sign(Data.Value-LP0.5))

  # zone0.val=with(stg.data,ifelse(zone==0,zs.s[1]+2*(LP3-Data.Value),NA))
  # zone9.val=with(stg.data,ifelse(zone==9,zs.s[9]+2*(Data.Value-UP2),NA))
  # zone6.val=with(stg.data,ifelse(zone==6,0,NA))
  # zone.val=with(stg.data,ifelse(zone%in%c(0,6,9),NA,
  #                 zs.s[ifelse(zone==0,1,zone)]+(zs.s[zone+1]-zs.s[ifelse(zone==0,1,zone)])*(Data.Value-score1)/(score2-score1)))





  if(allinfo==TRUE){
  rslt=stg.data[,c("Date","Data.Value","zone","LP0.5","score1","score2","penalty")]}else{
    rslt=stg.data[,c("Date","Data.Value","penalty")]
  }
  return(rslt)
}
